### PR TITLE
feat: Wrap rrweb callbacks in wrapper to better handle errors

### PR DESCRIPTION
**Pull request by mydea**
This adds a `callbackWrapper` method which we use to wrap any (async) callbacks in the record/observer codebase.
This wrapper will automatically try to add an `__rrweb__` property to any error thrown in the given callback.

Later in sentry, we can use this to determine to ignore a given error.

Note: I had to rewrite the monkey-patched CSSStyleSheet stuff into proxies, as otherwise I couldn't get the this-binding correctly via the prototype stuff. However, I think this is a nicer implementation anyhow, so should be good IMHO.

I also added some tests (this took >50% of the implementation time 🙈 as you can't easily access error objects from puppeteer) to verify this works.

The tests cover:

* `CSSStyleSheet` monkey patching
* monkey patching of `CSSGroupingRule`
* errors in mutation observer

This should give us a decent certainty that this generally works as expected. There _may_ be a place or two that I missed, but I don't think so. I tried to put the wrappers in a way that it encapsulates any sync stuff going on.

I'd ask you @billyvg & @Lms24 to check this out and look at it in the editor, to really see the flow of functions & if I didn't miss any relevant function call that could slip through.

cc @mitsuhiko

### CODE REVIEW

File: packages/rrweb/test/utils.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ import * as fs from 'fs';
 
 export async function launchPuppeteer() {
   return await puppeteer.launch({
-    headless: process.env.PUPPETEER_HEADLESS ? true : false,
+    headless: process.env.PUPPETEER_DEBUG ? false : true,
```
Comment (mydea) [inline packages/rrweb/test/utils.ts:20]: I switched this to default to headless, as actually I got some flaky tests locally without this. IMHO headless makes sense as default, unless you really want to  debug the chrome instance. If you think that's a bad change, I can revert this!

File: packages/rrweb/src/record/index.ts [hunk 2]
```diff
@@ -27,6 +27,7 @@ import {
 import { IframeManager } from './iframe-manager';
 import { ShadowDomManager } from './shadow-dom-manager';
 import { CanvasManager } from './observers/canvas/canvas-manager';
+import { callbackWrapper } from '../sentry/callbackWrapper';
```
Comment (mitsuhiko) [inline packages/rrweb/src/record/index.ts:30]: I think this should be generally upstreamable, so I would try to name this in a way that is not sentry specific.

**Review by mitsuhiko — CHANGES_REQUESTED** [review]

File: packages/rrweb/src/record/index.ts (hunk 2 above)
Comment (mitsuhiko) [reply packages/rrweb/src/record/index.ts:30]: Actually disregard this. I can see we need to separately upstream this with the amount of pending upstream changes already.

**Review by mitsuhiko — APPROVED** [review]

File: packages/rrweb/src/record/index.ts (hunk 2 above)
Comment (mydea) [reply packages/rrweb/src/record/index.ts:30]: Yeah, I'm actually thinking to upstream this as a more generic configurable `errorHandler` kind of thing. IMHO that probably makes the most sense from a generic rrweb usage perspective! Then we can configure the error handler to add the `__rrweb__` stuff in sentry-javascript for us (long term, when we eventually update our fork to v2)

**Comment by mydea** [comment]
ref https://github.com/rrweb-io/rrweb/pull/1107 - Upstream, I made a PR to add an `errorHandler` option, which we could use to handle this (if/when we update our fork to latest upstream)

File: packages/rrweb/test/utils.ts (hunk 1 above)
Comment (billyvg) [reply packages/rrweb/test/utils.ts:20]: for the flaky local tests, I've had better success with `runInBand`

**Review by Lms24 — APPROVED** [review]
I went through the PR in VSCode and tried to wrap my head around how these callbacks are are interconnected/nested. 
image

I'm afraid chances are pretty high that we missed _some_ weird async callback/edge case but I think overall it looks good and we now have a simple way of quickly wrapping any CB we might have missed. I guess the biggest challenge is figuring out that the individual, currently unwrapped CB are okay not to be wrapped because they're executed in sync in other CBs that are wrapped. 

Great that you upstreamed the error handler PR. If it's accepted I think this is a good reason to consider going to 2.x. Also noted this in https://github.com/getsentry/sentry-javascript/issues/6655 so we don't lose track of it.

File: packages/rrweb/src/record/observer.ts [hunk 3]
```diff
@@ -510,31 +515,44 @@ function initStyleSheetObserver(
...
-    const id = mirror.getId(this.ownerNode as INode);
-    if (id !== -1) {
-      styleSheetRuleCb({
-        id,
-        adds: [{ rule, index }],
-      });
-    }
-    return insertRule.apply(this, arguments);
-  };
+  win.CSSStyleSheet.prototype.insertRule = new Proxy(insertRule, {
```
[the remark this answers is not in the document]
Comment (mydea) [reply packages/rrweb/src/record/observer.ts:518]: Yeah, I think this is more of an additional safeguard. IMHO it doesn't hurt to have them (and having them be proxies is probably safer as well).

File: packages/rrweb/src/record/observer.ts [hunk 4]
```diff
@@ -672,29 +733,32 @@ function initMediaInteractionObserver({
...
     on('play', handler(MediaInteractions.Play)),
     on('pause', handler(MediaInteractions.Pause)),
     on('seeked', handler(MediaInteractions.Seeked)),
     on('volumechange', handler(MediaInteractions.VolumeChange)),
   ];
-  return () => {
+  return callbackWrapper(() => {
     handlers.forEach((h) => h());
-  };
+  });
```
[the remark this answers is not in the document]
Comment (mydea) [reply packages/rrweb/src/record/observer.ts:761]: As the forEach is called sync inside of the `callbackWrapper`, I think it is not necessary there!

**Pull request merged**